src/backend/cuda/blas.cpp

```diff
@@ -150,14 +156,116 @@ BLAS_FUNC(dot, cdouble, false, Z, u)
 #undef BLAS_FUNC
 #undef BLAS_FUNC_DEF
 
-using std::max;
-using std::vector;
+template<typename T>
+cudaDataType_t getType();
+
+template<>
+cudaDataType_t getType<float>() {
+    return CUDA_R_32F;
+}
+
+template<>
+cudaDataType_t getType<cfloat>() {
+    return CUDA_C_32F;
+}
+
+template<>
+cudaDataType_t getType<double>() {
+    return CUDA_R_64F;
+}
+
+template<>
+cudaDataType_t getType<cdouble>() {
+    return CUDA_C_64F;
+}
+
+template<>
+cudaDataType_t getType<half>() {
+    return CUDA_R_16F;
+}
+
+template<typename T>
+cublasGemmAlgo_t selectGEMMAlgorithm() {
+    auto dev              = getDeviceProp(getActiveDeviceId());
+    cublasGemmAlgo_t algo = CUBLAS_GEMM_DEFAULT;
+    return algo;
+}
+
+template<>
+cublasGemmAlgo_t selectGEMMAlgorithm<__half>() {
+    auto dev              = getDeviceProp(getActiveDeviceId());
+    cublasGemmAlgo_t algo = CUBLAS_GEMM_DEFAULT;
+    if (dev.major >= 7) { algo = CUBLAS_GEMM_DEFAULT_TENSOR_OP; }
+    return algo;
+}
 
 template<typename T>
-void gemm(Array<T> &out, af_mat_prop optLhs, af_mat_prop optRhs,
-          const T *alpha,
-          const Array<T> &lhs, const Array<T> &rhs,
-          const T *beta) {
+cublasStatus_t gemm_dispatch(BlasHandle handle, cublasOperation_t lOpts,
+                             cublasOperation_t rOpts, int M, int N, int K,
+                             const T *alpha, const Array<T> &lhs, dim_t lStride,
+                             const Array<T> &rhs, dim_t rStride, const T *beta,
+                             Array<T> &out, dim_t oleading) {
+    auto prop = getDeviceProp(getActiveDeviceId());
+    if (prop.major > 3) {
+        return cublasGemmEx(
+            blasHandle(), lOpts, rOpts, M, N, K, alpha, lhs.get(), getType<T>(),
+            lStride, rhs.get(), getType<T>(), rStride, beta, out.get(),
+            getType<T>(), out.strides()[1],
+            getType<T>(),  // Compute type
+
+            // NOTE: When using the CUBLAS_GEMM_DEFAULT_TENSOR_OP algorithm
+            // for the cublasGemm*Ex functions, the performance of the
+            // fp32 numbers seem to increase dramatically. Their numerical
+            // accuracy is also different compared to regular gemm fuctions.
+            // The CUBLAS_GEMM_DEFAULT algorithm selection does not experience
+            // this change. Does this imply that the TENSOR_OP function
+            // performs the computation in fp16 bit even when the compute
+            // type is CUDA_R_32F?
+            selectGEMMAlgorithm<T>());
+    } else {
+        using Nt = typename common::kernel_type<T>::native;
+        return gemm_func<Nt>()(blasHandle(), lOpts, rOpts, M, N, K, (Nt *)alpha,
+                               (Nt *)lhs.get(), lStride, (Nt *)rhs.get(),
+                               rStride, (Nt *)beta, (Nt *)out.get(), oleading);
+    }
+}
+
+template<typename T>
+cublasStatus_t gemmBatched_dispatch(BlasHandle handle,
```

Comment:
I don't like this mixed naming convention. `gemmBatched_dispatch` should be `gemmBatchedDispatch` and the other one should be `gemmDispatch`.